### PR TITLE
161967141 Add Search Functionality and Custom filtering based on parameters

### PR DESCRIPTION
**Pull request by jillo-abdullahi**
**What does this PR do?**
-  Implement search functionality and custom filtering based on parameters.

**Description of Task to be completed?**
- Add article search by title and author using keywords.
- Implement filters for title and author on all articles
- Implement filtering of articles using tags.

**How should this be manually tested?**

- Clone the repo and cd into the project folder
- Create a virtual environment by running `python3 -m venv <name_of_environment>`
- Activate the virtual environment by running `source <name_of_environment>/bin/activate`
- Run `$ pip install -r requirements.txt`
- Then add the following environment variables to your computer:

 - `$ export DB_NAME='choose_a_dbname'`
 - `$ export DB_USER='db_username'`
 - `$ export DB_PASSWORD='db_role_password'`
 - `$ export DB_HOST='localhost or preffered host'`
 - `$ export DB_PORT='server_port_number'`
- `$ export SECRET_KEY='some_secret_key`
- `$ export SOCIAL_AUTH_FACEBOOK_KEY='<facebook_app_id>'`
- `$ export SOCIAL_AUTH_FACEBOOK_SECRET='<facebook_app_secret>'`
- `$ export SOCIAL_AUTH_TWITTER_KEY='<twitter_api_key>'`
- `$ export SOCIAL_AUTH_TWITTER_SECRET='<twitter_api_secret>'`
- `$ export SOCIAL_AUTH_GOOGLE_OAUTH2_KEY='<google_client_id>'`
- `$ export SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET='<google_client_secret>'`

- Change the current branch by running `git checkout ft-search-filter-161967141`
- Run `python3 manage.py makemigrations` then, `python3 manage.py migrate`
- Run tests with `python3 manage.py test`
- Run the server with `python3 manage.py runserver`

**What are the relevant pivotal tracker stories?**
- [#161967141](https://www.pivotaltracker.com/story/show/161967141)

**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

**Screenshots**

screenshot 2018-12-20 at 14 25 46

screenshot 2018-12-20 at 14 30 22

screenshot 2018-12-20 at 14 29 35

screenshot 2018-12-20 at 14 32 50

### CODE REVIEW

**Comment by jillo-abdullahi** [comment]
@mutoromike @bettblake08 @Bryan-Cee @denomash @Darius-Ndubi @davishooly Hi team, you must have noticed that this PR was closed and then re-opened. I fixed several conflicts with the base branch locally but ended up with a Psycopg2 connection error when running tests whose source I could not trace. I opted to delete my local branch and re-create it off develop. I then added my original files and pushed to the remote branch. I did not expect this to close the PR as the remote branch was untouched. But the PR is now live and you can go ahead and review it. Thanks.

**Comment by Bryan-Cee** [comment]
I believe it would be good to have absolute imports than relative ones

**Comment by Bryan-Cee** [comment]
Update your screenshots on the PR.

**Review by Bryan-Cee — CHANGES_REQUESTED** [review]
Please check on all your imports and make them absolute for best practice.

**Comment by jillo-abdullahi** [comment]
@Bryan-Cee I have updated the screenshots. Thanks.

**Review by Bryan-Cee — APPROVED** [review]

**Comment by jillo-abdullahi** [comment]
@Bryan-Cee I have updated the relevant imports to absolute. Thanks for the recommendation.

File: authors/apps/articles/views/filters.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    serializer_class = GetArticlesSerializer
+
+    def get_queryset(self):
+        queryset = Article.objects.all()
+        tags = self.request.query_params.get('tags', '')
+        tag_list = tags.split(",")
+
+        if tag_list:
+            return queryset.filter(tag_list__name__contains=tag_list[0])
```
Comment (bettblake08) [inline authors/apps/articles/views/filters.py]: Can you change this line of code to accommodate a list of tags?

**Review by bettblake08 — CHANGES_REQUESTED** [review]

**Review by Darius-Ndubi — APPROVED** [review]
👍

File: authors/apps/articles/views/filters.py (hunk 1 above)
Comment (jillo-abdullahi) [reply authors/apps/articles/views/filters.py]: @bettblake08 This has been updated. Thanks

**Review by bettblake08 — APPROVED** [review]

**Review by mutoromike — APPROVED** [review]
LGTM

**Pull request merged**